Fix index out of bounds bug

msackman opened the pull request:
PutUint64 requires the slice to already have sufficient len - i.e. it writes to the slice, not appends. Previously, in this branch, the slice was not being extended, so PutUint64 would panic.

glycerine commented:
Thanks @msackman. Please add a test for the problem that this bug fixes.

msackman commented:
No, I won't add a test. To do so would require me to figure out how writePtr works and the semantics of everything it touches, which in turn would require me to probably learn the entire codebase and capnproto spec, which I do not have the time to do right now.

Given this bug, it is quite obvious that that particular branch has _never_ been run before.

msackman commented:
Having a further look around, https://github.com/glycerine/go-capnproto/blob/master/capn.go#L1439 contains exactly the same mistake. Anything that is `PutUint64(slice[len(slice):], foo)` cannot possibly ever work, and has never been reached.

Just to be clear, I'm not suggesting a test isn't necessary or desirable - it certainly is. But it's clearly difficult to hit these cases with integration-style tests - I've been using capnp every day for the last 6 months and only today have I hit these bugs. I'm only saying I do not have time right now to figure out how to hit these cases in integration tests - to do so would require knowledge of how the whole library works IMO.

zombiezen commented:
AFAICT, go-capnproto2 doesn't suffer from this problem. https://github.com/zombiezen/go-capnproto2/blob/master/capn.go#L337  However, I don't have tests to prove it; I've been meaning to write a test suite for the copying functionality.

glycerine commented:
@msackman Hi Matthew,  If you could post a snippet, however incomplete, that would be helpful.
if you can, could you describe what your code was doing that caused that branch to be taken? 
My stuff generally sticks to a single segment, as golang makes that easy. It would be quite useful to understand the specifics of how you were using the library.  In general I would not say that the multiple segment handling code is well tested or used. As Ross points out, you may find that his v2 re-write fixes many of these issues.

msackman commented:
@glycerine The weird thing is that I only saw this bug yesterday, and the only thing I've changed is that I've added crypto to all my TCP streams. So for capnp, the only change is that before, I was using `ReadFromStream`, whereas now I'm doing the reading + decryption myself and then using `ReadFromMemoryZeroCopy`. The top of the stack is:

    panic: runtime error: index out of range
    
    goroutine 240 [running]:
    github.com/glycerine/go-capnproto.(*Segment).writePtr(0xc820134f30, 0x20, 0xc820cab380, 0x0, 0x14, 0x1, 0x0, 0x2, 0x0, 0x0, ...)
    	/home/matthew/src/Go_external/src/github.com/glycerine/go-capnproto/capn.go:1423 +0x949
    github.com/glycerine/go-capnproto.Struct.SetObject(0xc820134f30, 0x18, 0x0, 0x8, 0x2, 0x1, 0x0, 0x0, 0xc820cab380, 0x0, ...)
    	/home/matthew/src/Go_external/src/github.com/glycerine/go-capnproto/capn.go:419 +0xc9

I'm afraid I can't go further in the stack at this point as it's not released just yet.

It's repeatable - if I disable the crypto then I never hit this bug. So it must have something to do with different sized buffers created by `ReadFromStream` versus what I'm doing with my crypto stuff.

In general, my product is a server and the response to a client may consist of gluing together various elements from all sorts of capnp-messages I've received in the past. So I'm constructing a new message to send back to a client, but populating the fields of that message with data from other messages. In this particular case though, it's _not_ this general pattern that is hitting the bug: I'm just setting a field which is a `Data` field and the `[]byte` that I'm setting it with is nothing special at all.

@zombiezen Thanks for the pointer to your work - I wasn't aware of that. Given I'm on the verge of launching my own work, I'm not going to switch right now; I've just run your tests looking at code coverage: it's a lot higher than here, which is great. Just wondering if you've considered using fuzz testers or similar? (please feel free to contact me directly - we probably shouldn't pollute this thread)
